### DISCUSSION TITLE
how to reference the real user id from realtime event hook

### DISCUSSION THREAD

**Discussion by SunChero** [pocketbase/pocketbase#1512] [Q&A] [answered]
hello,

im trying to build a user online presence by using the OnRealtimeConnectRequest event hooks .. except i cant find a way to retrieve the real user id behind the e.Client.Id()   which seem to be a unique Id for the connection ?

in my scenario , i would like to retrieve the user id and update a record attribute to show if the user is online/offline
```
app.OnRealtimeConnectRequest().Add(func(e *core.RealtimeConnectEvent) error {
		log.Println("connection event" + e.Client.Id())
		err = nc.Publish("chat.presence", []byte(fmt.Sprintf("client connection: %v", e.Client.Id())))
		if err != nil {
			log.Fatal(err)
		}
		//fmt.Println("Published message to subject:", subject)
		return nil
})
```

**Comment by ganigeorgiev** [comment, marked as answer]
You should be able to access the realtime connection auth record (if authenticated) by reading it from the `e.Client` store:

```go
authRecord, _ := e.Client.Get(apis.ContextAuthRecordKey).(*models.Record)
if authRecord != nil {
    log.Println(authRecord.Id)
}
```

**Comment by SunChero** [comment]
sorry where do you have  ContextAuthRecordKey coming from  ?
i tried something like 
```
authRecord, _ := e.HttpContext.Get(apis.ContextAuthRecordKey).(*models.Record)
fmt.Printf("connection event %v", authRecord)
````

i could see the struct data printed out but i could not access any function 
doing a authRecord.Get("attr")  result in a goroutine panic

im kinda stuck .. do you have a code example of this ?

Thank you

**Reply by ganigeorgiev** [reply]
> sorry where do you have ContextAuthRecordKey coming from ?

The constant is from the `apis` package. I forgot to prefix it, sorry.

For the realtime events you have to access the auth record from the connection client (aka. ` e.Client.Get(apis.ContextAuthRecordKey).(*models.Record)`), not from the `HttpContext`.

This is because the SSE authentication happens at a later stage, aka. after the initial SSE connection has been established (I'll document it better/add it in the hooks examples sometime later).

**Comment by SunChero** [comment]
humm somehow this is only working for disconnection events 
looks like the connection even is firing too early ?

this my code sample
app.OnRealtimeConnectRequest().Add(func(e *core.RealtimeConnectEvent) error {

		authRecord, _ := e.Client.Get(apis.ContextAuthRecordKey).(*models.Record)
		var user string
		if authRecord != nil {
			user = authRecord.GetString("profileId")
		} else {
			user = ""
		}
		fmt.Printf("user %v connected \n", user)
		return nil
	})
	app.OnRealtimeDisconnectRequest().Add(func(e *core.RealtimeDisconnectEvent) error {
		authRecord, _ := e.Client.Get(apis.ContextAuthRecordKey).(*models.Record)
		var user string
		if authRecord != nil {
			user = authRecord.GetString("profileId")
		} else {
			user = ""
		}
		fmt.Printf("user %v disconnected \n", user)
		return nil
	})
result:
```
> Server started at: http://127.0.0.1:8090
  - REST API: http://127.0.0.1:8090/api/
  - Admin UI: http://127.0.0.1:8090/_/
**### **user  connected**
2023/01/06 14:29:31 Realtime connection established: cbmqrNQcbfKuhivCtbz2XQACkA1JUsbXnObf6iX2**
[0.00ms] SELECT {{_collections}}.* FROM `_collections` WHERE [[id]] = '_pb_users_auth_' OR LOWER([[name]])='_pb_users_auth_' LIMIT 1
[0.00ms] SELECT `users`.* FROM `users` WHERE `users`.`id`='lx2grldh6hg47gc' LIMIT 1
[0.00ms] SELECT {{_collections}}.* FROM `_collections` WHERE [[id]] = '_pb_users_auth_' OR LOWER([[name]])='_pb_users_auth_' LIMIT 1
[0.00ms] SELECT `users`.* FROM `users` WHERE `users`.`id`='1q0slcb6uayatkw' LIMIT 1
[0.00ms] SELECT {{_collections}}.* FROM `_collections` WHERE [[id]] = 'profiles' OR LOWER([[name]])='profiles' LIMIT 1
[0.00ms] SELECT {{_collections}}.* FROM `_collections` WHERE [[id]] = 'woprisjr2g2ww3y' OR LOWER([[name]])='woprisjr2g2ww3y' LIMIT 1
[0.00ms] SELECT `profiles`.* FROM `profiles` WHERE `profiles`.`id`='oqg7g7c9uwbceau' LIMIT 1
### **user  connected**
2023/01/06 14:29:44 Realtime connection established: xYINg6UcOajK4Vi5kGyz3XQbLfCSzayg2bXfpZTa
[0.00ms] SELECT {{_collections}}.* FROM `_collections` WHERE [[id]] = '_pb_users_auth_' OR LOWER([[name]])='_pb_users_auth_' LIMIT 1
[0.00ms] SELECT `users`.* FROM `users` WHERE `users`.`id`='1q0slcb6uayatkw' LIMIT 1
2023/01/06 14:29:52 Realtime connection closed (cancelled request): xYINg6UcOajK4Vi5kGyz3XQbLfCSzayg2bXfpZTa
### **user oqg7g7c9uwbceau disconnected**
```

**Reply by ganigeorgiev** [reply]
I'll try to document it better. The above behavior is correct, because the SSE authentication happens at a later stage (during the subscription submission).

Or in other words, you should be able to access the auth record if you attach to any other [realtime event hook](https://pocketbase.io/docs/event-hooks/#realtime-api-hooks), for example:

```go
app.OnRealtimeAfterSubscribeRequest().Add(func(e *core.RealtimeSubscribeEvent) error {
	authRecord, _ := e.Client.Get(apis.ContextAuthRecordKey).(*models.Record)

	var user string
	if authRecord != nil {
		user = authRecord.GetString("profileId")
	}

	fmt.Printf("user %v subscribed \n", user)

	return nil
})
```

**Reply by SunChero** [reply]
thank you very much .. 
i dont know how it going to work with user presence reset or updated every new subscription request .. but it may actually work ..hopefully performance wont suffer a lot.

**Comment by itsSilver** [comment]
is it possible to do the same in `onRealtimeDisconnectRequest ` using js? i can get the `client.id()` but i need also the id of the user to update some fields after disconnecting

**Reply by ganigeorgiev** [reply]
Yes it should be possible as long as the client has submitted a subscription as authenticated user while the connection was active.

```js
onRealtimeDisconnectRequest((e) => {
    const authRecord = e.client.get("authRecord")

    console.log("user:", authRecord?.id)
})
```

**Reply by itsSilver** [reply]
thanks
